### PR TITLE
Narrative Generator refactoring

### PR DESCRIPTION
**Pull request by jamesagnew**
This PR reworks the narrative generator framework to allow narrative templates to be applied based on FHIRPath expressions (e.g. "this template only applies to Observation.where(code in 'http://loinc|123-3')"). It also splits out the "registry" of templates into a new separate class called "NarrativeTemplateManifest" so that we can incorporate the liquid stuff when it's ready.

@fil512 review please

### CODE REVIEW

File: hapi-fhir-structures-dstu3/src/test/java/ca/uhn/fhir/narrative2/ThymeleafNarrativeGeneratorTest.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+		gen.generateNarrative(composition);
+
+		// Firt narrative should be empty
+		String narrative = composition.getSection().get(0).getText().getDiv().getValueAsString();
+		assertThat(narrative, Matchers.emptyOrNullString());
+
+		// Second narrative should have details
+		narrative = composition.getSection().get(1).getText().getDiv().getValueAsString();
+		ourLog.info("Narrative:\n{}", narrative);
```
Comment (fil512) [inline hapi-fhir-structures-dstu3/src/test/java/ca/uhn/fhir/narrative2/ThymeleafNarrativeGeneratorTest.java]: Assert some details

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/model/base/composite/BaseResourceReferenceDt.java [hunk 2]
```diff
@@ -128,8 +128,9 @@ public BaseResourceReferenceDt setReference(IIdType theReference) {
 	}
 
 	@Override
-	public void setResource(IBaseResource theResource) {
+	public BaseResourceReferenceDt setResource(IBaseResource theResource) {
 		myResource = theResource;
+		return null;
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/model/base/composite/BaseResourceReferenceDt.java]: whoa, dum dum dum! good catch

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/BaseNarrativeGenerator.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.commons.lang3.StringUtils.defaultIfEmpty;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+public abstract class BaseNarrativeGenerator implements INarrativeGenerator {
+
+	private INarrativeTemplateManifest myManifest;
+	private FhirContext myFhirContext;
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/BaseNarrativeGenerator.java]: ah yeah, that should probably be set in a constructor argument as well. done.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/BaseNarrativeGenerator.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		} else {
+			return false;
+		}
+	}
+
+	private Optional<INarrativeTemplate> getTemplateForElement(IBase theElement) {
+		return myManifest.getTemplateByElement(getStyle(), theElement);
+	}
+
+	private boolean applyTemplate(INarrativeTemplate theTemplate, IBaseResource theResource) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/BaseNarrativeGenerator.java]: makes sense, done

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/BaseNarrativeGenerator.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		return myFhirContext;
+	}
+
+	public void setFhirContext(FhirContext theFhirContext) {
+		Validate.notNull(theFhirContext, "theFhirContext must not be null");
+		myFhirContext = theFhirContext;
+	}
+
+	@Override
+	public boolean generateNarrative(IBaseResource theResource) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/BaseNarrativeGenerator.java]: Fair point. How about `populateResourceNarrative(IBaseResource)`

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/NarrativeTemplate.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+		return myTemplateFileName;
+	}
+
+	void setTemplateFileName(String theTemplateFileName) {
+		myTemplateFileName = theTemplateFileName;
+	}
+
+	@Override
+	public Set<String> getAppliesToProfiles() {
+		return myAppliesToProfiles;
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/NarrativeTemplate.java]: you'll never see me complain about returning immutable structures!

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/ThymeleafNarrativeGenerator.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	protected String applyTemplate(INarrativeTemplate theTemplate, IBase theTargetContext) {
+
+		Context context = new Context();
+		context.setVariable("resource", theTargetContext);
+		context.setVariable("context", theTargetContext);
+		context.setVariable("fhirVersion", getFhirContext().getVersion().getVersion().name());
+
+		String result = getTemplateEngine().process(theTemplate.getTemplateName(), context);
+
+//		if (myCleanWhitespace) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/narrative2/ThymeleafNarrativeGenerator.java]: done

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java [hunk 8]
```diff
@@ -2350,6 +2350,31 @@ public void testSearchLinkToken() {
...
+		List<String> queries = CaptureQueriesListener
+			.getLastNQueries()
+			.stream()
+			.map(t -> t.getSql(true, true))
+			.collect(Collectors.toList());
+
+		ourLog.info("Queries:\n  {}", queries.stream().findFirst());
+
+		String searchQuery = queries.get(0);
+		assertEquals(searchQuery, 3, StringUtils.countMatches(searchQuery.toUpperCase(), "HFJ_SPIDX_TOKEN"));
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java]: haha i was proud of it :)

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/interceptor/api/Pointcut.java [hunk 9]
```diff
@@ -330,7 +335,31 @@
...
+
+	/**
+	 * Invoked when a resource may be returned to the user, whether as a part of a READ,
+	 * a SEARCH, or even as the response to a CREATE/UPDATE, etc.
+	 * <p>
+	 * This hook is invoked when a resource has been loaded by the storage engine and
+	 * is being returned to the HTTP stack for response. This is not a guarantee that the
+	 * client will ultimately see it, since filters/headers/etc may affect what
+	 * is returned but if a resource is loaded it is likely to be used.
+	 * </p>
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/interceptor/api/Pointcut.java]: done

File: hapi-fhir-structures-dstu3/src/test/java/ca/uhn/fhir/narrative2/ThymeleafNarrativeGeneratorTest.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-structures-dstu3/src/test/java/ca/uhn/fhir/narrative2/ThymeleafNarrativeGeneratorTest.java]: Done.

**Comment by lawley** [comment]
really excited about this one

**Pull request merged**